examples/vae_simple.py

```diff
@@ -0,0 +1,319 @@
+                    name = 'dec_mu')
+            # relu_shift is for numerical stability - if training data has any
+            # dimensions where stdev=0, allowing logsigma to approach -inf
+            # will cause the loss function to become NAN. So we set the limit
+            # stdev >= exp(-1 * relu_shift)
+            relu_shift = 10
+            l_dec_logsigma = nn.layers.DenseLayer(l_dec_hid, num_units=x_dim,
+                    W = nn.init.GlorotUniform() if W_dec_ls is None else W_dec_ls,
+                    b = nn.init.Constant(0) if b_dec_ls is None else b_dec_ls,
+                    nonlinearity = lambda a: T.nnet.relu(a+relu_shift)-relu_shift,
+                    name='dec_logsigma')
+            l_output = GaussianSampleLayer(l_dec_mu, l_dec_logsigma,
+                    name='dec_output')
+            l_dec_mu_list.append(l_dec_mu)
+            l_dec_logsigma_list.append(l_dec_logsigma)
+            l_output_list.append(l_output)
+            if W_dec_hid is None:
+                W_dec_hid = l_dec_hid.W
+                b_dec_hid = l_dec_hid.b
+                W_dec_mu = l_dec_mu.W
+                b_dec_mu = l_dec_mu.b
+                W_dec_ls = l_dec_logsigma.W
+                b_dec_ls = l_dec_logsigma.b
+        l_Z_list.append(l_Z)
+    l_output = nn.layers.ElemwiseSumLayer(l_output_list, coeffs=1./L, name='output')
+    return l_enc_mu, l_enc_logsigma, l_dec_mu_list, l_dec_logsigma_list, l_output_list, l_output
+
+# ############################## Main program ################################
+
+# helper function for log-likelihood expression
+def log_likelihood(tgt, mu, ls):
+    return T.sum(-(np.float32(0.5 * np.log(2 * np.pi)) + ls)
+            - 0.5 * T.sqr(tgt - mu) / T.exp(2 * ls))
+
+def main(L=2, z_dim=2, n_hid=1024, num_epochs=300, binary=True):
+    print("Loading data...")
+    X_train, X_val, X_test = load_dataset()
+    width, height = X_train.shape[2], X_train.shape[3]
+    input_var = T.tensor4('inputs')
+
+    # Create VAE model
+    print("Building model and compiling functions...")
+    print("L = {}, z_dim = {}, n_hid = {}, binary={}".format(L, z_dim, n_hid, binary))
+    x_dim = width * height
+    l_z_mu, l_z_ls, l_x_mu_list, l_x_ls_list, l_x_list, l_x = \
+           build_vae(input_var, L=L, binary=binary, z_dim=z_dim, n_hid=n_hid)
+
+    # If there are dropout layers etc these functions return masked or non-masked expressions
+    # depending on if they will be used for training or validation/test err calcs
+    layer_outputs = lambda b: nn.layers.get_output([l_z_mu, l_z_ls] +
+            l_x_mu_list + l_x_ls_list + l_x_list, deterministic=b)
+    z_mu = lambda b: layer_outputs(b)[0]
+    z_ls = lambda b: layer_outputs(b)[1]
+    x_mu = lambda b: [] if binary else layer_outputs(b)[2:2+L]
+    x_ls = lambda b: [] if binary else layer_outputs(b)[2+L:2+2*L]
+    x_list = lambda b: layer_outputs(b)[2 if binary else 2+2*L:]
```

Comment:
Sorry, but this still doesn't do a single `get_output()` call... `z_mu` and `z_ls` (and everything else) calls `layer_outputs`, which calls `get_output()`. I think you should just define a local `build_loss(deterministic)` function that does everything without lambdas, then call that twice below. Something like:
```python
def build_loss(deterministic):
    layer_outputs = nn.layers.get_output([l_x, ...] + ... + ... + ..., deterministic=deterministic)
    x = layer_outputs[0]
    z_mu = layer_outputs[1]
    z_ls = layer_outputs[2]
    ...
    kl_div = ...
    if binary:
        logpxz = ...
        prediction = x   # note: avoid another get_output() call, in case you ever want to compute the output and the loss at the same time
    else:
        logpxz = ...
        prediction = ...
    return -1 * (logpxz + kl_div), prediction

loss, _ = build_loss(deterministic=False)
test_loss, test_prediction = build_loss(deterministic=True)
```